### PR TITLE
Ksagiyam/merge upstream

### PR DESCRIPTION
**Pull request by ksagiyam**
Update `petsc` and `slepc`.

Must land with https://github.com/florianwechsung/TinyASM/pull/10 for tests to pass.

### CODE REVIEW

File: .github/workflows/build.yml [hunk 1]
```diff
@@ -46,7 +46,7 @@ jobs:
       - name: Build Firedrake
         run: |
           cd ..
-          ./firedrake/scripts/firedrake-install $COMPLEX --venv-name build --tinyasm --disable-ssh --minimal-petsc --slepc --documentation-dependencies --install thetis --install gusto --install icepack --install irksome --install femlium --no-package-manager|| (cat firedrake-install.log && /bin/false)
+          ./firedrake/scripts/firedrake-install $COMPLEX --venv-name build --tinyasm --disable-ssh --minimal-petsc --slepc --documentation-dependencies --install thetis --install gusto --install icepack --install irksome --install femlium --no-package-manager --package-branch petsc ksagiyam/merge_upstream --package-branch slepc ksagiyam/merge_upstream|| (cat firedrake-install.log && /bin/false)
```
Comment (JDBetteridge) [inline .github/workflows/build.yml:49]: Don't forget to remove these when PETSc/SLEPc are updated

File: scripts/firedrake-install [hunk 2]
```diff
@@ -1240,6 +1240,9 @@ def build_and_install_tinyasm():
             # Clean source directory
             check_call(["git", "reset", "--hard"])
             check_call(["git", "clean", "-f", "-x", "-d"])
+            check_call(["git", "remote", "add", "temp", "https://github.com/francesco-ballarin/TinyASM.git"])
```
Comment (JDBetteridge) [inline scripts/firedrake-install:1243]: Don't forget to update!

**Review by JDBetteridge — APPROVED** [review]
Ping me when PETSc/SLEPc update goes through and I'll re-approve (or use your superpowers).

**Comment by ksagiyam** [comment]
Thanks. I will drop branches and push.

Test results with updated petsc/slepc/TinyASM are here:

https://github.com/firedrakeproject/firedrake/actions/runs/2797849393

**Comment by ksagiyam** [comment]
This PR turned out to be harmless, so I just merge.

**Comment by ksagiyam** [comment]
The following could be useful when GAMG breaks Firedrake in the future:

PETSc MFE:
https://gitlab.com/petsc/petsc/-/issues/1221

Firedrake equivalent (almost):
```
from firedrake import *
import numpy as np


n = 64
mesh = UnitSquareMesh(n, n)
V = VectorFunctionSpace(mesh, "CG", 2)
u = TrialFunction(V)
v = TestFunction(V)
x, y = SpatialCoordinate(mesh)
inside_box = And(abs(x-0.5) < 0.2, abs(y-0.75) < 0.2)
mu = conditional(inside_box, 1e8, 1)


a = inner(mu * sym(grad(u)), grad(v)) * dx
f = as_vector((0, -9.8*n*n*conditional(inside_box, 2, 1)))
L = inner(f, v) * dx
bcs = [DirichletBC(V.sub(0), 0., (1, 2, 3, 4)), DirichletBC(V.sub(1), 0., (1, 2, 3, 4))]


rotV = Function(V).interpolate(as_vector((-y, x)))
c0V = Function(V).interpolate(Constant([1., 0.]))
c1V = Function(V).interpolate(Constant([0., 1.]))
near_nullmodes = VectorSpaceBasis([c0V, c1V, rotV])
near_nullmodes.orthonormalize()

solver_parameters = {
    'ksp_type': 'cg',
    'pc_type': 'python',
    'pc_python_type': 'firedrake.AssembledPC',
    'assembled_pc_type': 'gamg',
    'assembled_mg_levels_pc_type': 'sor',
    'assembled_mg_levels_pc_sor_diagonal_shift': 1e-100,
    'ksp_rtol': 1e-7,
    'ksp_converged_reason': None,
    'ksp_view': None,
}
A = assemble(a)
w = Function(V)
problem = LinearVariationalProblem(a, L, w, bcs=bcs)
solver = LinearVariationalSolver(
    problem,
    near_nullspace=near_nullmodes,
    solver_parameters=solver_parameters)
solver.solve()
print(np.linalg.norm(w.dat.data, ord=np.inf))
```

**Pull request merged**